src/py21cmfast/drivers/param_config.py

```diff
@@ -0,0 +1,249 @@
+                    "We are currently unable to use PERTURB_ON_HIGH_RES and HALO_STOCHASTICITY"
+                    "Simultaneously."
+                )
+                raise NotImplementedError(msg)
+
+        if val.USE_EXP_FILTER and not val.USE_HALO_FIELD:
+            warnings.warn("USE_EXP_FILTER has no effect unless USE_HALO_FIELD is true")
+
+    @astro_params.validator
+    def _astro_params_validator(self, att, val):
+        if val is None:
+            return
+        if self.user_params is None:
+            return
+
+        if val.R_BUBBLE_MAX > self.user_params.BOX_LEN:
+            raise InputCrossValidationError(
+                f"R_BUBBLE_MAX is larger than BOX_LEN ({val.R_BUBBLE_MAX} > {self.user_params.BOX_LEN}). This is not allowed."
+            )
+
+        if (
+            global_params.HII_FILTER == 1
+            and val.R_BUBBLE_MAX > self.user_params.BOX_LEN / 3
+        ):
+            msg = (
+                "Your R_BUBBLE_MAX is > BOX_LEN/3 "
+                f"({val.R_BUBBLE_MAX} > {self.user_params.BOX_LEN / 3})."
+            )
+
+            if config["ignore_R_BUBBLE_MAX_error"]:
+                warnings.warn(msg)
+            else:
+                raise ValueError(msg)
+
+    def merge_keys(self):
+        """The list of available structs in this instance."""
+        # Allow using **input_parameters
+        return [
+            field.name
+            for field in attrs.fields(self.__class__)
+            if field.name != "redshift"
+        ]
+
+    def __getitem__(self, key):
+        """Get an item from the instance in a dict-like manner."""
+        # Also allow using **input_parameters
+        return getattr(self, key)
+
+    def is_compatible_with(self, other: InputParameters) -> bool:
+        """Check if this object is compatible with another parameter struct.
+
+        Compatibility is slightly different from strict equality. Compatibility requires
+        that if a parameter struct *exists* on the other object, then it must be equal
+        to this one. That is, if astro_params is None on the other InputParameter object,
+        then this one can have astro_params as NOT None, and it will still be
+        compatible. However the inverse is not true -- if this one has astro_params as
+        None, then all others must have it as None as well.
+        """
+        if not isinstance(other, InputParameters):
+            return False
+
+        return not any(
+            other[key] is not None and self[key] is not None and self[key] != other[key]
+            for key in self.merge_keys()
+        )
+
+    def merge(self, other: InputParameters) -> InputParameters:
+        """Merge another InputParameters instance with this one, checking for compatibility."""
+        if not self.is_compatible_with(other):
+            raise ValueError("Input parameters are not compatible.")
+        return InputParameters(**{k: self[k] or other[k] for k in self.merge_keys()})
```

Comment:
Noting here that if a user manually merges two structs, redshift is set to None. This forces the user to manually specify redshifts before using the InputParamters struct to initialise an OutputStruct

Reply:
Our normal behvaiour is usually to only merge using `from_output_structs` which asks for a redshift. and subsequently verify using `check_redshift_consistency` with the outputs we want to be consistent with